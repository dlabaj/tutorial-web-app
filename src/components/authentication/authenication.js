import React from 'react';
import PropTypes from 'prop-types';
import { connect, reduxActions } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import { noop, OC_MODE } from '../../common/helpers';

class Authentication extends React.Component {
  state = {
    email: '',
    emailError: null,
    password: '',
    passwordError: null,
    remember: false,
    formTouched: false,
    formValid: false
  };

  componentDidMount() {
    const { session, checkUser, storeData } = this.props;

    if (!session.authorized) {
      checkUser();
      storeData();
    }
  }

  onChangeRemember = event => {
    const { checked } = event.target;
    const { removeStoredData } = this.props;

    if (!checked) {
      removeStoredData().then(() =>
        this.setState({
          remember: checked
        })
      );
    } else {
      this.setState({
        remember: checked
      });
    }
  };

  onLogin = event => {
    const { email, password, remember, formValid } = this.state;
    const { checkUser, loginUser, removeStoredData, storeData } = this.props;

    event.preventDefault();

    if (formValid) {
      this.setState(
        {
          formTouched: false,
          password: '',
          passwordError: null
        },
        () =>
          loginUser({
            [apiTypes.API_SUBMIT_AUTH_USERNAME]: email,
            [apiTypes.API_SUBMIT_AUTH_PASSWORD]: password
          }).then(() => {
            if (remember) {
              storeData({ email });
            } else {
              this.setState(
                {
                  email: '',
                  emailError: null
                },
                () => removeStoredData()
              );
            }

            return checkUser();
          })
      );
    }
  };

  static getDerivedStateFromProps(props, state) {
    let updateInitialState = null;

    if (OC_MODE) {
      updateInitialState = {
        email: (props.session.remember && props.session.storedEmail) || state.email,
        password: state.password,
        emailError: '',
        passwordError: '',
        formValid: true
      };
    } else if (!state.formTouched && props.session.remember && props.session.storedEmail) {
      updateInitialState = {
        email: props.session.storedEmail,
        emailError: '',
        remember: props.session.remember
      };
    }

    return updateInitialState;
  }

  isFormValid() {
    const { emailError, passwordError } = this.state;
    const formValid = emailError === '' && passwordError === '';

    this.setState({
      formValid
    });
  }

  render() {
    const { children, session } = this.props;
    if (session.authorized) {
      return children;
    }

    return null;
  }
}

Authentication.propTypes = {
  checkUser: PropTypes.func,
  children: PropTypes.node.isRequired,
  loginUser: PropTypes.func,
  removeStoredData: PropTypes.func,
  session: PropTypes.shape({
    error: PropTypes.bool,
    loginFailed: PropTypes.bool,
    authorized: PropTypes.bool,
    pending: PropTypes.bool,
    storedEmail: PropTypes.string,
    remember: PropTypes.bool
  }),
  storeData: PropTypes.func
};

Authentication.defaultProps = {
  checkUser: noop,
  loginUser: noop,
  removeStoredData: noop,
  session: {},
  storeData: noop
};

const mapDispatchToProps = dispatch => ({
  checkUser: () => dispatch(reduxActions.userActions.checkUser()),
  loginUser: data => dispatch(reduxActions.userActions.loginUser(data)),
  storeData: data => dispatch(reduxActions.userActions.storeData(data)),
  removeStoredData: () => dispatch(reduxActions.userActions.removeStoredData())
});

const mapStateToProps = state => ({ session: state.userReducers.session });

const ConnectedAuthentication = connect(
  mapStateToProps,
  mapDispatchToProps
)(Authentication);

export { ConnectedAuthentication as default, ConnectedAuthentication, Authentication };
