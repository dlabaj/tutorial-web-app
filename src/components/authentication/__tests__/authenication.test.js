import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import { ConnectedAuthentication, Authentication } from '../authenication';

describe('Authentication Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a basic component with login form', () => {
    const store = generateEmptyStore({
      userReducers: { session: { error: false, loginFailed: false, pending: false } }
    });
    const component = shallow(
      <ConnectedAuthentication>
        <span className="test">lorem</span>
      </ConnectedAuthentication>,
      { context: { store } }
    );

    expect(component).toMatchSnapshot('connected');
  });

  it('should render a basic component without login form', () => {
    const props = {
      session: {
        authorized: true
      }
    };

    const component = mount(
      <Authentication {...props}>
        <span className="test">lorem</span>
      </Authentication>
    );

    expect(component.render()).toMatchSnapshot('post-authorization');
  });
});
