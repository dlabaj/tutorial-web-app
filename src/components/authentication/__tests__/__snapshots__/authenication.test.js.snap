// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Component should have specific events defined: expected state 1`] = `
Object {
  "email": "",
  "emailError": "Email must be valid",
  "formTouched": true,
  "formValid": false,
  "password": "123",
  "passwordError": "",
  "remember": true,
}
`;

exports[`Authentication Component should render a basic component with login form: connected 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Authentication Component should render a basic component without login form: post-authorization 1`] = `
<span
  class="test"
>
  lorem
</span>
`;
